### DISCUSSION TITLE
Syncing to external sync signal vs internal timers

### DISCUSSION THREAD

**Discussion by richms** [Roger-random/ESP_8_BIT_composite#19] [Ideas]
I have used a modified TV out library made for the nootropic design arduino shield in the past to let me have multiple video signals in sync with each other. It was made for their shield that would overlay video ontop of a signal, but it also meant that the output video was in sync with the signal so that when the crt it was displayed on was being videoed it would not have a rolling bar as the clocks drifted.

Anyway, its all good and stuff but this has colour, which is a huge win over monochrome video. I dont know enough about programming to know if it would be possible to take the hsync and vsync from the nootropic shield into a couple of pins of the ESP32 and have it use those as a timing source rather than the internal timers.

**Comment by Roger-random** [comment]
I don't know anything about nootropic shield so I don't know what their hsync and vsync signals look like.

If someone is familiar with that information and wants to give this idea a try, here are some pointers:

For horizontal sync, ```ESP_8_BIT_composite.cpp``` has the following global variables: 
```
_hsync
_hsync_long
_hsync_short
```
The current implementation configures them once never modifies them again. A modification to use external hsync timing will need to modify code that interacts with these variables.

For vertical sync, the ```video_isr()``` function defined in ```ESP_8_BIT_composite.cpp``` knows when one frame ends and starts the next frame. At the moment there is a bit of code to swap front and back buffers (Look for comment ```// frame is done```) and this is likely the key location to implement any kind of external vsync capacibility.

**Comment by marciot** [comment]
I believe at least half of this request would be easy. For the vertical sync, I suspect a little bit of code could be added to the `video_isr` to check the vertical sync pin. If that was high, it would simply reset `_line_counter` to zero. As a proof of concept, you could probably change [the following line](https://github.com/Roger-random/ESP_8_BIT_composite/blob/a232df2517d60aa07a5fa1491914c91336c8e6d8/ESP_8_BIT_composite.cpp#L620) to:

```
    if (_line_counter == _line_count || digitalRead(VSYNC_PIN)) {
```

From:

```
    if (_line_counter == _line_count) {
```

Now the horizontal sync I believe would be much harder. As far as I understand, the timing of horizontal lines is controlled by the Audio PLL in the ESP32 which writes out the memory buffer continuously and triggers the ISR at the start of each new line. There is *probably* some way to tell it to reset itself, so it can start writing a line just as a new hsync pulse come in, but I wouldn't even know where to start with that. My understanding of the stuff related to the Audio PLL is basically zero at this point.
